### PR TITLE
Update static graph spec

### PR DESCRIPTION
**Pull request by cdmihai**
With the details from #4218 and caching.

### CODE REVIEW

File: documentation/specs/static-graph.md [hunk 1]
```diff
@@ -237,16 +310,43 @@ There is also the possibility for these higher-order build engines and even Visu
...
+<!-- workflow -->
+Single project builds can be achieved by providing MSBuild with input and output cache files.
+
+The input cache files contain the cached results of all of the current project's references. This way, when the current project executes, it will naturally build its references via [MSBuild task](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-task) calls. The engine, instead of executing these tasks, will serve them from the provided input caches. 
+
+The output cache file tells MSBuild where it should serialize the results of the current project. This output cache would become an input cache for all other projects that depend on the current project.
+The output cache file can be ommited in which case the build would just reuse prior results but not write out any new results. This could be useful when one wants to replay a build from previous caches.
+
+The presence of either input or output caches turns on the isolated build constraints. The engine will fail the build on:
+- MSBuild task calls on projects files which were not defined in the `ProjectReference` item at evaluation time.
```
Comment (rainersigwald) [inline documentation/specs/static-graph.md]:
```suggestion
- `MSBuild` task calls to project files which were not defined in the `ProjectReference` item at evaluation time.
```

File: documentation/specs/static-graph.md [hunk 2]
```diff
@@ -237,16 +310,43 @@ There is also the possibility for these higher-order build engines and even Visu
...
+- MSBuild task calls which cannot be served from the cache
```
Comment (rainersigwald) [inline documentation/specs/static-graph.md]:
```suggestion
- `MSBuild` task calls which cannot be served from the cache
```

File: documentation/specs/static-graph.md [hunk 3]
```diff
@@ -237,16 +310,43 @@ There is also the possibility for these higher-order build engines and even Visu
...
+
+<!-- cache structure -->
+These cache files contain the serialized state of MSBuild's [ConfigCache](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/Components/Caching/ConfigCache.cs) and [ResultsCache](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/Components/Caching/ResultsCache.cs). These two caches have been traditionally used by the engine to cache build results.
+
+Cache structure: `(project path, global properties) -> results`
+
+<!-- cache lifetime -->
+The caches are applicable for the entire duration of the MSBuild.exe process. The input and output caches have the same lifetime as the `ConfigCache` and the `ResultsCache`. The `ConfigCache` and the `ResultsCache` are owned by the [BuildManager](https://github.com/Microsoft/msbuild/blob/master/src/Build/BackEnd/BuildManager/BuildManager.cs), and their lifetimes are one `BuildManager.BeginBuild` / `BuildManager.EndBuild` session. Since MSBuild.exe uses one BuildManager with one BeginBuild / EndBuild session, the cache lifetime is the same as the entire process lifetime.
```
Comment (rainersigwald) [inline documentation/specs/static-graph.md]: This gets a little wonky with node reuse, but I think it's ok to leave.

File: documentation/specs/static-graph.md [hunk 4]
```diff
@@ -237,16 +310,43 @@ There is also the possibility for these higher-order build engines and even Visu
...
+
+<!-- constraints -->
+The following are some constraints enforced by the engine. They're just for
```
Comment (rainersigwald) [inline documentation/specs/static-graph.md]: Just for?

File: documentation/specs/static-graph.md [hunk 5]
```diff
@@ -237,16 +310,43 @@ There is also the possibility for these higher-order build engines and even Visu
...
+
+<!-- constraints -->
+The following are some constraints enforced by the engine. They're just for 
 
-In fact, this same mechanism can be used for non-distributed builds as well, for example for incremental builds. The caller would still need to ensure that the build result cache was valid (ie. the project reference project was not changed since the build result cache for it was produced), but the benefit would be that only the current project being built would need to be evaluated and only targets on that project would need to execute.
+Input cache files constraints:
+- A ConfigCache / ResultsCache mapping must be unique between all input caches (multiple input caches cannot provide information for the same cache entry)
+- For each input cache, ConfigCache.Entries.Size == ResultsCache.Entries.Size
```
Comment (rainersigwald) [inline documentation/specs/static-graph.md]:
```suggestion
- For each input cache, `ConfigCache.Entries.Size == ResultsCache.Entries.Size`
```

File: src/Build.UnitTests/Graph/ResultCacheBasedBuilds_Tests.cs [hunk 6]
```diff
@@ -353,24 +391,26 @@ public void MissingResultFromCacheShouldErrorDueToIsolatedBuildCacheEnforcement(
             results["1"].Logger.Errors.First().Message.ShouldContain("MSB4252");
         }
 
+        // These method runs in two modes:
+        // populateDictionaries = true
```
Comment (rainersigwald) [inline src/Build.UnitTests/Graph/ResultCacheBasedBuilds_Tests.cs]: and?

File: src/Build/BackEnd/BuildManager/CacheSerialization.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+using System;
+using System.IO;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Execution
+{
+    internal static class CacheSerialization
+    {
+        public static string SerializeCaches(IConfigCache configCache, IResultsCache resultsCache, string outputCacheFile)
```
Comment (rainersigwald) [inline src/Build/BackEnd/BuildManager/CacheSerialization.cs]: Are there meaningful differences here or just a code move/cleanup?

**Comment by cdmihai** [comment]
@AndyGerlicher @erikma @dannyvv

File: documentation/specs/static-graph.md [hunk 8]
```diff
@@ -33,36 +35,71 @@
...
 
-For example if project B targeted `net471` and had a project reference to the example project A above, then project B would depend on project A (without global properties). However when building project B from the command-line, project A would only build its `net471` (project A + global properties of `TargetFramework=net471`) variant, and the `netcoreapp2.0` variant would not actually be built.
+For example, if project A had a project reference to project B with `GlobalPropertiesToRemove=Platform`, and we wanted to build project A for x86 and x64 so used both as entry points, the graph would consist of 3 nodes: project A with `Platform=x86`, project A with `Platform=x64`, and project B with no global properties set.
 
-Because we are using ProjectReferences to determine the graph, we will need to duplicate the mapping of ProjectReference metadata to global properties given to the MSBuild task. This means that we need to couple the engine to the common MSBuild targets, which means that whenever we change those targets we need to update the project graph construction code as well.
+#### Crosstargeting
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md]: So that's the official name? If so, then I will update the other PR too.

**Comment by cdmihai** [comment]
> mostly LGTM but I'm not sure I understand the code changes: are they integral to the doc change, or just more convenient to bundle?

Realized that the spec made an assertion for which no unit test existed, so decided to add the unit test. While adding the test, I need to expose extra stuff, and while exposing extra stuff I realized I can do a nice refactoring to decouple some stuff out of BuildManager.

File: documentation/specs/static-graph.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md]: Ah, but all the files are named *.Crosstargeting.*, so now there's ambiguity. Is this some official rename and current artifacts with `Crosstargeting` in their names are grandfathered in as "whatever"?

[the remark this answers is not in the document]
Comment (rainersigwald) [reply documentation/specs/static-graph.md]: Essentially, yeah.

File: documentation/specs/static-graph.md [hunk 9]
```diff
@@ -33,36 +35,71 @@
...
+<!-- how it works: outer builds and inner builds -->
+Crosstargeting is implemented by having a project reference itself multiple times, once for each combination of crosstargeting global properties. This leads to multiple evaluations of the same project, with different global properties. These evaluations can be classified in two groups
+1.  Multiple inner builds. Each inner build is evaluated with one set of crosstargeting global properties (e.g. the `TargetFramework=net472` inner build, or the `TargetFramework=netcoreapp2.2` inner build).
+2.  One outer build. This evaluation does not have any crosstargeting global properties set. It can be viewed as a proxy for the inner builds. Other projects query the outer build in order to learn the set of valid crosstargeting global properties (the set of valid inner builds). When the outer build is also the root of the project to project graph, the outer build multicasts the entry target (i.e. `Build`, `Clean`, etc) to all inner builds.
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md]: Not really, as `Build`, `Clean` are known graph wide operation for the SDK, and its multicasting code knows about both of them. Other SDKs that would want to implement their own multitargeting scheme, would have their own separate graph wide operations and their separate multicasting machinery. Unless the intent is that `Microsoft.Common.Crosstargeting.targets` should be used by all potentially crosstargeting SDKs. So I'd say to ignore this issue until the next crosstargeting SDK comes along and we have to actually decide on common infrastructure.

File: documentation/specs/static-graph.md [hunk 10]
```diff
@@ -33,36 +35,71 @@
...
+In order for the graph to represent inner and outer builds as nodes, it imposes a contract on what crosstargeting means, and requires the crosstargeting supporting SDKs to implement this contract.
 
-**OPEN ISSUE:** What about MSBuild calls which pass in global properties not on the ProjectReference? Worse, if the global properties are computed at runtime? The sfproj SDK does this with PublishDir, which tells the project where to put its outputs so the referencing sfproj can consume them. We may need to work with sfproj folks and other SDK owners, which may not be necessarily a bad thing since already the sfproj SDK requires 2 complete builds of dependencies (normal ProjectReference build + this PublishDir thing) and can be optimized. A possible short term fix: Add nodes to graph to special case these kinds of SDKs. This is only really a problem for isolated builds.
+Crosstargeting supporting SDKs need to implement the following properties and semantics:
+- `InnerBuildProperty`. It contains the property name that defines the crosstargeting build dimension.
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md]: I think the indirection should be there, to decouple crosstargeting from specific names.

File: documentation/specs/static-graph.md [hunk 11]
```diff
@@ -33,36 +35,71 @@
...
+- `InnerBuildPropertyValues`. It contains the property name that holds the possible values for the `InnerBuildProperty`.
+- Project classification:
+  - *Outer build*, when `$($(InnerBuildProperty))` is empty AND  `$($(InnerBuildPropertyValues))` is not empty.
+  - *Dependent inner build*, when both `$($(InnerBuildProperty))` and  `$($(InnerBuildPropertyValues))` are non empty. These are inner builds that were generated from an outer build.
+  - *Standalone inner build*, when `$($(InnerBuildProperty))` is not empty and `$($(InnerBuildPropertyValues))` is empty. These are inner builds that were not generated from an outer build.
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md]: `TargetFramework` is not filled in for legacy csprojes:
image

So then with these `non crosstargeting` projects, SDKs can allow a mix of crosstargeting and non crosstargeting projects. If the need arises, we can add extra settings on how these are interpreted, like whether the SDK allows non crosstargeting projects or not.

[the remark this answers is not in the document]
Comment (rainersigwald) [reply documentation/specs/static-graph.md]: I guess the larger question is: is there really a strong distinction between a single-targeted SDK-style project and a legacy csproj?

File: documentation/specs/static-graph.md [hunk 12]
```diff
@@ -33,36 +35,71 @@
...
+  - *Non crosstargeting build*, when both `$($(InnerBuildProperty))` and  `$($(InnerBuildPropertyValues))` are empty.
+- Node edges
+  - When project A references crosstargeting project B, and B is identified as an outer build, the graph node for project A will reference both the outer build of B, and all the inner builds of B. The edges to the inner builds are speculative, as at build time only one inner build gets referenced. However, the graph cannot know at evaluation time which inner build will get chosen.
```
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md]: For now, yes. In the (near?) future, these edges will have some extra metadata so API users can traverse speculative edges. One use is post-processing the graph to filter for only one entry TF. Which would be useful for inner-loop scenarios where devs just want to quickly iterate on one TF and not wait for everything.

File: documentation/specs/static-graph.md (hunk 3 above)
Comment (cdmihai) [reply documentation/specs/static-graph.md]: Should be fine with NodeReuse, as the nodes get "reimaged", so all cache results are deleted between reuses.

File: documentation/specs/static-graph.md (hunk 4 above)
Comment (cdmihai) [reply documentation/specs/static-graph.md]: Just for showing my sometimes short attention span :)

File: src/Build.UnitTests/Graph/ResultCacheBasedBuilds_Tests.cs (hunk 6 above)
Comment (cdmihai) [reply src/Build.UnitTests/Graph/ResultCacheBasedBuilds_Tests.cs]: Short attention span strikes again.

File: src/Build/BackEnd/BuildManager/CacheSerialization.cs (hunk 7 above)
Comment (cdmihai) [reply src/Build/BackEnd/BuildManager/CacheSerialization.cs]: Just moving two highly coupled methods into their own type.

File: documentation/specs/static-graph.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (cdmihai) [reply documentation/specs/static-graph.md]: For the current SDK not really, but the code needs to reason about each type of project, so every project needs a type :) It seemed somehow wrong to classify these non-participating projects as standalone inner builds, so I added this extra one. Let me know if you feel strongly about removing it.

**Pull request merged**